File: Battleships/Scripts/Typescript/Battleships.js
var Battleships;
(function (Battleships) {
    var BattleshipsGame = (function () {
        function BattleshipsGame() {
            var _this = this;
            this.takeShot = function (shotCoordinates) {
                if (!Grid.gridStringValid(shotCoordinates)) {
                    return "invalid coordinates";
                }
                var gridCell = Grid.parse(shotCoordinates);
                if (gridCell === null) {
                    return "cant generate cell";
                }
                if (_this.cellAlreadyShotAt(gridCell)) {
                    return "already shot at";
                }
                var shipHit = _this.returnShipHit(gridCell);
                if (shipHit === null) {
                    return "miss";
                }
                if (_this.wasShotFatalBlow(shipHit)) {
                    return "fatality!";
                }
                return "Hit on ship";
            };
            this.cellAlreadyShotAt = function (cell) {
                var existingShot = _this.shots.filter(function (s) { return Grid.areGridCellsEqual(s.coordinatesOfShot, cell); });
                if (existingShot.length > 0) {
                    return true;
                }
                return false;
            };
            this.returnShipHit = function (cell) {
                var shipHit = _this.returnIndexOfShipWhichHasComponentAtCoordinate(cell);
                if (shipHit === -1) {
                    _this.recordShot(cell, false);
                    return null;
                }
                _this.recordShot(cell, true);
                _this.setShipComponentToHit(shipHit, cell);
                return shipHit;
            };
            this.returnIndexOfShipWhichHasComponentAtCoordinate = function (cell) {
                for (var s = 0; s < _this.ships.length; s++) {
                    var ship = _this.ships[s];
                    var matchingCoordinates = ship.components.filter(function (c) { return Grid.areGridCellsEqual(c.coordinates, cell); });
                    if (matchingCoordinates.length > 0) {
                        return s;
                    }
                }
                return -1;
            };
            this.recordShot = function (cell, didHit) {
                _this.shots.push(new Shots.Shot(cell, didHit));
            };
            this.setShipComponentToHit = function (shipIndex, cell) {
                var ship = _this.ships[shipIndex];
                for (var s = 0; s < ship.components.length; s++) {
                    if (Grid.areGridCellsEqual(ship.components[s].coordinates, cell)) {
                        _this.ships[shipIndex].components[s].hasBeenHit = true;
                        break;
                    }
                }
            };
            this.wasShotFatalBlow = function (shipIndex) {
                var shipHit = _this.ships[shipIndex];
                return shipHit.components.every(function (c) { return c.hasBeenHit; });
            };
            this.allShipsDestroyed = function () {
                var shipsDestroyed = _this.ships.every(function (s) { return s.components.every(function (c) { return c.hasBeenHit; }); });
                console.log(shipsDestroyed);
                return shipsDestroyed;
            };
            var shipGenerator = new ShipGenerators.ShipGenerator();
            this.ships = shipGenerator.ships;
            this.shots = new Array();
        }
        return BattleshipsGame;
    })();
    Battleships.BattleshipsGame = BattleshipsGame;
})(Battleships || (Battleships = {}));

File: BattleshipsConsole/Battleships.cs
using System.Collections.Generic;
using System.Linq;

namespace BattleshipsConsole
{
    public class Battleships
    {
        private const string InvalidGridReference = "The grid reference you entered: {0} is invalid. A-J1-10 only.";
        private const string AlreadyShotMessage = "You have already taken a shot at {0}. Hit a ship: {1}.";
        private const string MissMessage = "Shot taken at {0}....... miss :(";
        private const string HitMessage = "Shot taken at {0}....... Its a direct hit! :D";
        private const string SunkMessage = "BOOM! the last remaining component of {0} has been hit. Status: Sunk! o_O";
        private readonly List<Ship> _ships;
        private readonly List<Shot> _shots;
        private readonly GridCellComparer _gridCellComparer;

        public Battleships()
        {
            var shipGenerator = new ShipGenerator();
            _ships = shipGenerator.Ships;
            _shots = new List<Shot>();
            _gridCellComparer = new GridCellComparer();
        }

        public string TakeShot(string enteredCell)
        {
            GridCell shotCoordinates;
            if (!GridCell.TryParse(enteredCell, out shotCoordinates))
                return string.Format(InvalidGridReference, enteredCell);

            if (CellHasAlreadyBeenShotAt(shotCoordinates))
                return GenerateAndReturnAlreadyShotMessage(shotCoordinates);

            Ship shipHit;
            if (!DidShotHitAShip(shotCoordinates, out shipHit))
                return string.Format(MissMessage, shotCoordinates);

            return WasShotTheFatalBlow(shipHit) ? 
                string.Format(SunkMessage, shipHit.Name) : 
                string.Format(HitMessage, shotCoordinates);
        }

        private bool CellHasAlreadyBeenShotAt(GridCell cellCoordinates)
        {
            return _shots.Any(s => _gridCellComparer.Equals(s.CoordinatesOfShot, cellCoordinates));
        }

        private string GenerateAndReturnAlreadyShotMessage(GridCell cell)
        {
            var shot = _shots.FirstOrDefault(s => _gridCellComparer.Equals(s.CoordinatesOfShot, cell));
            return string.Format(AlreadyShotMessage, cell, shot != null && shot.DidHit);
        }

        private bool DidShotHitAShip(GridCell shotCoordinates, out Ship shipHit)
        {
            shipHit = ReturnShipWhichHasComponentAtCoordinates(shotCoordinates);
            if (shipHit == null)
            {
                RecordShot(shotCoordinates, false);
                return false;
            }
            var componentHit = shipHit.Components.FirstOrDefault(c => _gridCellComparer.Equals(c.Coordinates, shotCoordinates));
            componentHit.HasBeenHit = true;
            RecordShot(shotCoordinates, true);
            return true;
        }

        private static bool WasShotTheFatalBlow(Ship ship)
        {
            return ship.Components.All(c => c.HasBeenHit);
        }

        private Ship ReturnShipWhichHasComponentAtCoordinates(GridCell cellCoordinates)
        {
            return _ships.FirstOrDefault(s => s.Components.Any(c => _gridCellComparer.Equals(c.Coordinates, cellCoordinates)));
        }

        private void RecordShot(GridCell cell, bool wasHit)
        {
            _shots.Add(new Shot(cell, wasHit));
        }

        public bool AllShipsDestroyed
        {
            get { return _ships.All(s => s.Components.All(c => c.HasBeenHit)); }
        }

        public int NoOfShotsTaken => _shots.Count;
    }
}
